flocker/node/agents/blockdevice.py

```diff
@@ -1442,234 +1626,93 @@ def _mountpath_for_dataset_id(self, dataset_id):
-        """
-        :param PMap devices: The datasets with volumes attached to this node
-            and the device files at which they are available.  This is the same
-            as ``NodeState.devices``.
-        :param PMap paths: The paths at which datasets' filesystems are mounted
-            on this node.  This is the same as ``NodeState.paths``.
-        :param PMap configured: The manifestations which are configured on this
-            node.  This is the same as ``NodeState.manifestations``.
-        :param volumes: An iterable of ``BlockDeviceVolume`` instances that are
-            known to exist in the cluster.
-
-        :return: A generator of ``DetachVolume`` instances, one for each
-            dataset which exists, is attached to this node, is not mounted, and
-            is configured to not have a manifestation on this node.
-        """
-        for attached_dataset_id in devices:
-            if unicode(attached_dataset_id) in configured:
-                # It is supposed to be here.
-                continue
-            if unicode(attached_dataset_id) in paths:
-                # It is mounted and needs to unmounted before it can be
-                # detached.
+        not_in_use_datasets = not_in_use(local_datasets.values())
+        for dataset_id, dataset in local_datasets.items():
+            if dataset in not_in_use_datasets:
                 continue
-            volume = _blockdevice_volume_from_datasetid(volumes,
-                                                        attached_dataset_id)
-            yield DetachVolume(dataset_id=attached_dataset_id,
-                               blockdevice_id=volume.blockdevice_id)
-
-    def _calculate_attaches(self, devices, configured, nonmanifest, volumes):
-        """
-        :param PMap devices: The datasets with volumes attached to this node
-            and the device files at which they are available.  This is the same
-            as ``NodeState.devices``.
-        :param PMap configured: The manifestations which are configured on this
-            node.  This is the same as ``NodeState.manifestations``.
-        :param PMap nonmanifest: The datasets which exist in the cluster but
-            are not attached to any node.
-        :param volumes: An iterable of ``BlockDeviceVolume`` instances that are
-            known to exist in the cluster.
-
-        :return: A generator of ``AttachVolume`` instances, one for each
-                 dataset which exists, is unattached, and is configured to be
-                 attached to this node.
-        """
-        for manifestation in configured.values():
-            dataset_id = UUID(manifestation.dataset_id)
-            if dataset_id in devices:
-                # It's already attached here.
+            if dataset.state not in (
+                DatasetStates.ATTACHED, DatasetStates.MOUNTED
+            ):
+                # A lease doesn't force a mount.
```

Comment:
Regarding `NON_MANIFEST`, I was going to comment that while it wasn't necessary (since missing `DesiredDataset`s get treated as `NON_MANIFEST`) it makes sense to have them explicit. But, we don't have any configuration for non-manifest nodes. I could loop through every other node and mark all the manifestations there, but I'll just leave a comment as to why we don't discover `NON_MANIFEST` right now.